File: app/workers/weather_poller_worker.rb
# Polls weather for subscribed locations
class WeatherPollerWorker
  include Sidekiq::Worker
  include Sidekiq::Repeat::Repeatable

  repeat { hourly.minute_of_hour(0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55) }

  def perform
    logger.info "Running weather poll"

    locations.each do |location|
      poll(location)
    end
  end

  def subscriptions
    $redis.pubsub("channels", "helios2_*:graphql-event::weatherPublished:*")
  end

  LocationParams = Struct.new(:latitude, :longitude)

  PARSE_LOCATION = /latitude:(?<latitude>.+):longitude:(?<longitude>.+)$/

  def locations
    subscriptions.map do |subscription|
      captures = PARSE_LOCATION.match(subscription).named_captures
      LocationParams.new(*captures.values)
    end.uniq
  end

  def poll(location)
    logger.info "Polling #{location.latitude} #{location.longitude}"
    location_record = Location.where(latitude: location.latitude, longitude: location.longitude).first
    get_forecast(location_record, location)
  end

  def get_forecast(record, location)
    data = Clients::DarkskyClient.forecast(record)
    Helios2Schema.subscriptions.trigger(
      "weatherPublished",
      { latitude: location.latitude.to_f, longitude: location.longitude.to_f },
      data
    )
  end
end

File: server-phoenix/lib/helios/workers/weather_poller_worker.ex
defmodule Helios.Workers.WeatherPollerWorker do
  use Absinthe.Schema.Notation
  import Ecto.Query
  alias HeliosWeb.Schema.Types.Sub
  alias HeliosWeb.Clients.WeatherClient

  def perform do
    IO.puts("Running weather poll")

    Enum.each(locations, fn location ->
      get_forecast(location)
    end)
  end

  def subscriptions do
    Sub.get_args() |> Enum.uniq()
  end

  defmodule LocationParams do
    defstruct [:latitude, :longitude]
  end

  def locations do
    Enum.map(subscriptions, fn subscription ->
      %LocationParams{latitude: subscription.latitude, longitude: subscription.longitude}
    end)
  end

  def get_forecast(location) do
    case WeatherClient.forecast(%{latitude: location.latitude, longitude: location.longitude}) do
      {:ok, forecast_data} ->
        Absinthe.Subscription.publish(
          HeliosWeb.Endpoint,
          forecast_data,
          weather_published: [location.latitude, location.longitude]
        )

      {:error, message} ->
        IO.puts("Failure polling #{location.latitude} and #{location.longitude}: #{message}")
    end
  end
end
